js/style/style.js

```diff
@@ -580,7 +580,11 @@ Style.prototype = util.inherit(Evented, {
     setPaintProperty: function(layerId, name, value, klass) {
         this._checkLoaded();
 
-        this.getLayer(layerId).setPaintProperty(name, value, klass);
+        var layer = this.getLayer(layerId);
+
+        if (layer.getPaintProperty(name, klass) === value) return this;
```

Comment:
Do you think it's worth doing a `deepEqual` comparison instead? `value` can be an object (function) or an array.